src/Stratis.Bitcoin/Utilities/AsyncQueue.cs

```diff
@@ -93,11 +113,95 @@ public void Enqueue(T item)
             }
         }
 
+        /// <summary>
+        /// Dequeues an item from the queue if there is one. 
+        /// If the queue is empty, the method waits until an item is available.
+        /// </summary>
+        /// <param name="cancellation">Cancellation token that allows aborting the wait if the queue is empty.</param>
+        /// <returns>Dequeued item from the queue.</returns>
+        /// <exception cref="OperationCanceledException">Thrown when the cancellation token is triggered or when the queue is disposed.</exception>
+        /// <exception cref="InvalidOperationException">Thrown if this method is called on a queue that operates in callback mode.</exception>
+        public async Task<T> DequeueAsync(CancellationToken cancellation = default(CancellationToken))
+        {
+            if (this.callbackMode)
+                throw new InvalidOperationException($"{nameof(DequeueAsync)} called on queue in callback mode.");
+
+            // Increment the counter so that the queue's cancellation source is not disposed when we are using it.
+            Interlocked.Increment(ref this.unfinishedDequeueCount);
+
+            try
+            {
+                if (this.disposed)
+                    throw new OperationCanceledException();
+
+                // First check if an item is available. If it is, just return it.
+                T item;
+                if (this.TryDequeue(out item))
+                    return item;
+
+                // If the queue is empty, we need to wait until there is an item available.
+                using (var cancellationSource = CancellationTokenSource.CreateLinkedTokenSource(cancellation, this.cancellationTokenSource.Token))
+                {
+                    while (true)
+                    {
+                        await this.signal.WaitAsync(cancellationSource.Token).ConfigureAwait(false);
```

Comment:
If cancellation is called this will throw an Operation Cancellation exception right?
then the `Interlocked.Decrement(ref this.unfinishedDequeueCount);` will decrement by one, later bellow this might block the dispose code that uses this
```
 while (this.unfinishedDequeueCount > 0)
                Thread.Sleep(5);
```

Reply:
yes cancellation will cause this to throw
then yes, decrement is called
why would this block dispose?

Reply:
If cancellation throws you will decrements the counter only by one, what if there are more items in the queue?

Reply:
the decrement is for the number of pending dequeue operations (think threads), so when you enter dequeue operation you increase by 1, when you finish it, you decrease by one

this try-finally makes sure that every increase has its decrease

it has nothing to do with the content of the queue

Reply:
Ah my bad I did not read the logic correctly, the counter represents dequeuing threads not items.

Brilliant.

I would suggest a rename in this case `unfinishedDequeueCount` to `listenerCount`

Reply:
it is weird because i chose such a not so nice name "unfinishedDequeueCount" in order to be so clear that it is really the number of unfinished dequeue operations :-) and its comment was supposed to make it even more clear 

listenerCount to me sounds not so explicit

Reply:
hmm I didnt read the documentation thoroughly.
How ever `unfinishedDequeueCount` is feels like its talking about items, if anything then name it `unfinishedDequeue[r]Count`

But I think listener does explain exactly the what this count is for, how many threads (or callers) are currently waiting in the `unfinishedDequeueCount` counter.
and ListenerCount sounds better then DequeuerCount.

However this is really semantics if you don't think the name change is more correct let push it as is.